tests/testutils/goroutineassistant.go

```diff
@@ -46,14 +50,49 @@ func (a *GoroutineAssistant) Done() {
 }
 
 func (a *GoroutineAssistant) Wait() {
-	done := make(chan struct{})
 	go func() {
 		a.wg.Wait()
-		done <- struct{}{}
+		a.s <- nil
 	}()
-	select {
-	case s := <-a.s:
-		a.t.Fatalf(s)
-	case <-done:
+	err := <-a.s
+	if err == nil {
+		// success
+		return
+	}
+	// If we received an error, let's fatal with that one. But for clean
+	// test teardown, we need to allow the other goroutines to shut down.
+	// We log any other errors we encounter in the meantime.
+	a.t.Logf("Error encountered - shutting down")
+	defer a.t.Fatal(err)
+	teardown := time.After(5 * time.Minute)
```

Comment:
5 minutes? Ew. Also, no need for the `teardown` variable - just use `case <-time.After(some_timeout)` in `select` below.